(WIP) lazy log messages: Replacing or continuing #105

acodili-jg opened the pull request:
### Changes from #105:

#### Removals

 - `boolean ALUtils.checkErrors(Supplier<String>)`

#### Additions

 - `Optional<String> ALUtils.consumeErrorMessage()`
 - `boolean ALUtils.logAndConsumeApplyError(String, int, String, int)`
 - `boolean ALUtils.logAndConsumeError(Object)`
 - `boolean ALUtils.logAndConsumeMultiApplyError(String[], int[], String, int)`
 - `boolean ALUtils.logAndConsumeSetError(String, String, String, float)`
 - `boolean ALUtils.logAndConsumeSetError(String, String, int, float)`
 - `boolean ALUtils.logAndConsumeSetError(String, int, String, float)`

#### Changes

 - `boolean ALUtils.errorApply(String, int, String, int)`
    - delegates to `boolean ALUtils.logAndConsumeApplyError(String, int, String, int)`
 - `boolean ALUtils.errorApply(String[], int[], String, int)`
    - delegates to `boolean ALUtils.logAndConsumeMultiApplyError(String[], int[], String, int)`
 - `boolean ALUtils.errorProperty(String, String, int, float)`
    - delegates to `boolean ALUtils.logAndConsumeSetError(String, String, int, float)`
 - `boolean ALUtils.errorProperty(String, int, String, float)`
    - delegates to `boolean ALUtils.logAndConsumeSetError(String, int, String, float)`
 - `boolean Cache.generate(Consumer<String>)`
    - changed signature to `boolean Cache.generate(LogBuilder)`
       - all member calls to the first parameter are replaced with an equivalent
    - the recent `boolean Cache.generate()` might supersede this method

<details>
<summary>Old content</summary>

 - ~~`checkErrors(Supplier<String>)` is renamed to `logErrors(Supplier<String>)`~~
    - ~~The rename better describes the method, though it seems that `ALUtils.checkErrors(String)` is based on Minecraft's `ALUtil.checkErrors(String)`~~
 - Replaced `checkErrors(Supplier<String>)`
 - Revert changes to `checkErrors(String)` and wrapped the string concat in a supplier
 - Rename Cache#generate(Consumer<String>) to Cache#generate(LogBuilder) ([2b7bf56](2b7bf5608874a56b40cc2408db64ad1b8f1c0965))
 - Replaced `Consumer<String>` loggers to `LogBuilder`, usually acquired using `Logger.atXxx()` or `Logger.atLevel(Level)`, loggers
    - Though `LogBuilder` is more similar to `Logger`, it could be more restrictive than `Consumer<String>` (e.g. `message -> { doSomething(message); LOGGER.log(message); }`).
 - Updated commented code
 - Some templated logging message lines' parameters are wrapped into a supplier
 - Replace streamed for-each logging into a multiline message sent in bulk
    - Benefit is that the stream and the operations are only performed when logging can happen (e.g. matching log level)
    - Behavioral changes, for a theoretical scenario where the logging level changes mid-iterating the stream:
       - Previous behavior would partially print the entries of the stream
       - Current implementation would either print the whole stream or none of it
</details>
 - and some more...

acodili-jg committed:
#105, but the branch is based on the octree one

checkErrors(Supplier<String>) is renamed to logErrors with the same signature. New overload with logging level parameter.

acodili-jg committed:
Remove logErrors(Level, Supplier<String>)

acodili-jg committed:
ALUtils#checkErrors(String) - wrap log message concat in a supplier

acodili-jg committed:
Update Cache#generate(Consumer<String>) to Cache#generate(LogBuilder)

Changed the parameter type `Consumer<String>` of *logger* with `LogBuilder`, usually acquired using `Logger.atXxx()` or `Logger.atLevel(Level)`, as it functions nearly the same as an actual logger (e.g.: formatting, lazy messages, etc.).

acodili-jg committed:
Engine: Update logging stuff

 - Update to calls to `Cache.generate(Consumer<String>)` to `Cache.generate(LogBuilder)`
 - Inline the local variable `startTime` in `playSound(Context, vec3d, int, boolean)` to make it `final`
 - Made a multiple line string to a template multiline string for lazy logging
 - Changed the type `Consumer<String>` of the local variable `logger` to `LogBuilder` and updated related code.

acodili-jg committed:
Updated commented logging code

acodili-jg committed:
Update PrecomputedConfig.java

Simplified 2nd log message formatting parameter and instead directly reference the local variable `wrong`

acodili-jg committed:
Cache.generate(LogBuilder): Update streamed logging to be sent in bulk (lazily)

acodili-jg committed:
Merge branch 'octree-testing' into wip-lazy-log-messages

acodili-jg committed:
Missed a logging line in Effect.java

acodili-jg committed:
Merge branch 'wip-lazy-log-messages' of https://github.com/acodili-jg/resounding into wip-lazy-log-messages

acodili-jg committed:
ALUtils.checkAlcErrors(long, String): Wrapped logged message template parameters in a supplier

acodili-jg committed:
Context.(de)activate(): Changed calls to ALUtils.checkErrors(String) to ALUtils.logErrors(Supplier<String>)

acodili-jg committed:
Outdated local origin, oopsies...

mikenrafter commented on src/main/java/dev/thedocruby/resounding/Cache.java:
```diff
@@ -282,17 +285,17 @@ public static Branch growOctree(WorldChunk chunk, Branch root) {
+                    .map(RegistryKey::getValue)
+                    .map(Identifier::getPath)
+                    .collect(Collectors.joining("\n"));
+        });
```
While this code isn't of much consequence, given it's being refactored... Multiple maps have worse performance than a single iteration, and this degrades readability in this circumstance.

acodili-jg replied:
I see. I will replace this with a single map or maybe find an alternative to streamTags().

mikenrafter reviewed (changes requested):
I haven't fully tested these changes, but most all seems well!

mikenrafter commented:
I'll merge this once I get the cache system built.

acodili-jg committed:
Collect the multiple maps and joining collector into a single collect

 - I couldn't resist naming it like that...
 - Explanation: I checked how `Collectors.joinning(charSequence)` was 'implemented', made the code to be based-ish on that, and finally merge the maps into the 2nd parameter "accumulator"
 - Yeah, I haven't found a way around `Registry.streamTags()`

acodili-jg committed:
SoundSystemMixin: Update commented code

 - Similar changes to `9404647fb82cf4f6478c8de155d2419dfae1f82e`

acodili-jg committed:
Merge branch 'octree-testing' into wip-lazy-log-messages

acodili-jg committed:
Cache: Remove an exta empty line from the imports

mikenrafter commented:
How are you addressing: [lambdas on the heap](https://stackoverflow.com/questions/27524445/does-a-lambda-expression-create-an-object-on-the-heap-every-time-its-executed) and [lambda closure](https://www.infoq.com/articles/Java-8-Lambdas-A-Peek-Under-the-Hood/)?

Do you think performance degradation when logging is worth it? If so, could we get some profiling done to know what kind of hit these closure-accessing lambdas cause?

\**not being condescending, genuinely curious to hear your take*

mikenrafter committed:
Merge branch 'octree-testing' into wip-lazy-log-messages

acodili-jg committed:
Engine: "Problematic" unnoticed System.nanoTime() in lambda-d logging

Reverted lambda-fication for time logging.

acodili-jg commented:
These didn't come to my mind before being mentioned. Though most lambda-*fications* are mostly done to `ALUtils` (and code calling to its members), I was ~~over~~ thinking about limiting the number of lambda closures. Right now, I'm currently thinking about making an alternative to (not outright replace) `ALUtils.checkErrors(Consumer<String>)` as I'm aiming to use parameterized message logging to replace most lambdas that are connected to `ALUtils`.

<details>
<summary></summary>

> I had no access to the internet for two days. I did get to see this PR for a bit on that last day. This message totally didn't take a whole morning to reword.
</details>

acodili-jg committed:
Update ALUtils.java

`getAlcErrorMessage(int)` wasn't coslty to begin with.

mikenrafter committed:
Merge branch 'octree-testing' into wip-lazy-log-messages

acodili-jg committed:
ALUtils: Log and consume methods

acodili-jg committed:
Replace lingering references to logErrors with alternatives

acodili-jg committed:
Update Engine.java

Replace ternary with if statement for startTime

acodili-jg committed:
ALUtils: Readded prefix to logAndConsumeMultiApplyError's logging

acodili-jg committed:
Merge branch 'wip-lazy-log-messages' of https://github.com/acodili-jg/resounding into wip-lazy-log-messages

acodili-jg committed:
PrecomputedConfig: Fixed wrong to be wrapped in [], similar to pre-PR

The original was `Arrays.toString(new List[]{wrong})`, which should result to `[[actual wrong list contents here]]`. Before this commit, the new changes unwrapped it by one; this commit fixes that.

acodili-jg committed:
Update commented logging code

acodili-jg committed:
Merge branch 'octree-testing' into wip-lazy-log-messages

acodili-jg committed:
Engine: Add missing imports

acodili-jg committed:
Engine: Typo `LGGER`

acodili-jg committed:
ModClient: Update commented code

acodili-jg committed:
SoundSystemMixin: Update commented code

acodili-jg committed:
Merge branch 'octree-testing' into wip-lazy-log-messages

StartsMercury commented:
Inactive and outdated. I haven't checked in a while but I think it is better to create a new PR if this is still needed (and maybe start over since these aren't big changes) and close this one for now.

acodili-jg commented:
Inactive and outdated. I haven't checked in a while but I think it is better to create a new PR if this is still needed (and maybe start over since these aren't big changes) and close this one for now.
